project/SparkVersionAxis.scala

```diff
@@ -32,27 +33,22 @@ object SparkVersionAxis {
       .toLowerCase()
   }
 
-  implicit class ProjectExtension(val p: ProjectMatrix) extends AnyVal {
+  implicit class ProjectExtension(val projectMatrix: ProjectMatrix) extends AnyVal {
 
-    def sparkRow(sparkAxis: SparkVersionAxis, scalaVersions: Seq[String], settings: Def.SettingsDefinition*): ProjectMatrix =
-      p.customRow(
-        scalaVersions = scalaVersions,
-        axisValues = Seq(sparkAxis, VirtualAxis.jvm),
-        _.settings(
+    def sparkRow(sparkAxis: SparkVersionAxis, scalaVersions: Seq[String], settings: Def.SettingsDefinition*): ProjectMatrix = {
+      val sparkVersion = sparkAxis.sparkVersion
+      scalaVersions.foldLeft(projectMatrix)((currentProjectMatrix, scalaVersion) =>
+        currentProjectMatrix.customRow(
+          scalaVersions = Seq(scalaVersion),
+          axisValues = Seq(sparkAxis, VirtualAxis.jvm),
+          _.settings(
             moduleName := camelCaseToLowerDashCase(name.value + sparkAxis.directorySuffix),
-            libraryDependencies ++= dependencies(sparkAxis.sparkVersion)
-        ).settings(settings: _*)
-      )
-
-    def sparkRow(sparkAxis: SparkVersionAxis, scalaVersion: String, settings: Def.SettingsDefinition*)
-                (implicit jacocoReportSettings: JacocoReportSettings): ProjectMatrix =
-      p.customRow(
-        scalaVersions = Seq(scalaVersion),
-        axisValues = Seq(sparkAxis, VirtualAxis.jvm),
-        _.settings(
-          moduleName := camelCaseToLowerDashCase(name.value + sparkAxis.directorySuffix),
-          libraryDependencies ++= dependencies(sparkAxis.sparkVersion)
-        ).settings(settings: _*)
+            libraryDependencies ++= dependencies(sparkAxis.sparkVersion),
+            jacocoReportSettings := jacocoSettings(sparkVersion, scalaVersion),
```

Comment:
Just curious here: Why has the settings moved here -- in the project matrix? 

I can see that the spark/scala version is now to be included in the title, but is that it?
Is and was Jacoco run for each spark/scala combination individually and has this changed somehow?

Reply:
This location is the first place in "our" project where exact combination of version is known. So it is used to correctly set each variation of JaCoCo report.